streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -16,10 +16,294 @@
+
+            // As we go, check each topology for overlap in the set of input topics/patterns
+            final int numInputTopics = allInputTopics.size();
+            final List<String> inputTopics = builder.fullSourceTopicNames();
+            final Collection<String> inputPatterns = builder.allSourcePatternStrings();
+
+            final int numNewInputTopics = inputTopics.size() + inputPatterns.size();
+            allInputTopics.addAll(inputTopics);
+            allInputTopics.addAll(inputPatterns);
+            if (allInputTopics.size() != numInputTopics + numNewInputTopics) {
+                inputTopics.retainAll(allInputTopics);
+                inputPatterns.retainAll(allInputTopics);
+                inputTopics.addAll(inputPatterns);
+                log.error("Tried to add the NamedTopology {} but it had overlap with other input topics: {}", builder.namedTopology(), inputTopics);
+                throw new TopologyException("Named Topologies may not subscribe to the same input topics or patterns");
+            }
+
+            final ProcessorTopology globalTopology = builder.buildGlobalStateTopology();
+            if (globalTopology != null) {
+                if (builder.namedTopology() != null) {
+                    throw new IllegalStateException("Global state stores are not supported with Named Topologies");
+                } else if (this.globalTopology == null) {
+                    this.globalTopology = globalTopology;
+                } else {
+                    throw new IllegalStateException("Topology builder had global state, but global topology has already been set");
+                }
+            }
+            globalStateStores.putAll(builder.globalStateStores());
+        });
+    }
+
+    public ProcessorTopology buildSubtopology(final TaskId task) {
+        return lookupBuilderForTask(task).buildSubtopology(task.subtopology());
+    }
+
+    public ProcessorTopology globalTaskTopology() {
+        if (hasNamedTopologies()) {
+            throw new IllegalStateException("Global state stores are not supported with Named Topologies");
+        }
+        return globalTopology;
+    }
+
+    public Map<String, StateStore> globalStateStores() {
+        return globalStateStores;
+    }
+
+    public Map<String, List<String>> stateStoreNameToSourceTopics() {
+        final Map<String, List<String>> stateStoreNameToSourceTopics = new HashMap<>();
+        applyToEachBuilder(b -> stateStoreNameToSourceTopics.putAll(b.stateStoreNameToSourceTopics()));
+        return stateStoreNameToSourceTopics;
+    }
+
+    public String getStoreForChangelogTopic(final String topicName) {
+        for (final InternalTopologyBuilder builder : builders.values()) {
+            final String store = builder.getStoreForChangelogTopic(topicName);
+            if (store != null) {
+                return store;
+            }
+        }
+        log.warn("Unable to locate any store for topic {}", topicName);
+        return "";
```

Comment:
Why not return `null` here?

Reply:
Just being consistent with what was returned in that case before. If we return `null` then an NPE would be thrown atm. 

It's used by `KafkaStreams#allLocalStorePartitionLags` ie KIP-535. Could definitely be cleaned up, but I don't want to pull that into scope of this PR